### DISCUSSION TITLE
Modelling a TLV (type length value)

### DISCUSSION THREAD

**Discussion by ckrenslehner** [jam1garner/binrw#290] [Q&A] [answered]
Hi! 

I have a very general question. How would you model a TLV? I am using binrw quite a lot and I want to learn about `idiomatic` usage of this amazing crate. :-)

The basic structure should be as follows: `type (u8) ; length(u16) ; value`

Particularly I want to model `value` as an enum but still want to be able to calculate the `length` similarly to the way I would calculate it, if `value` was a vector.

Here is a snippet for further illustration:

``` rust
#[binrw]
#[br(import(r#type: u8))]
enum Value {
    #[br(assert(r#type == 1))]
    Temperature(f32),
    #[br(assert(r#type == 2))]
    Opened(
        #[br(map = |x: u8| x != 0)]
        #[bw(map = |x: &bool| if *x { 1 } else { 0 })]
        bool,
    ),
}

#[binrw]
struct TLV {
    r#type: u8,
    /// This value should be calculated by using the encoded size of `value`.
    /// If `value` was a `Vec<u8>` I could use `calc` to write the correct size.
    length: u16,
    #[br(args(r#type))]
    value: Value,
}
```

When reading a `TLV` it would be nice to validate, that the size encoded size of `value` indeed exactly matches the value of `length`. How would you tackle this aspect?
I am thinking of `TakeSeek` or using a custom parser function.

Ah and btw how do you model the encoding of `bool`?

Thanks and greetings,
Christian

**Comment by csnover** [comment, marked as answer]
> Here is a snippet for further illustration:

Use `pre_assert`, not `assert`, for selecting an enum variant. For unknown types, specify a fallback variant as the last variant with no `pre_assert` with `#[br(count = length)] Vec<u8>` to collect data for an unknown type.

> When reading a TLV it would be nice to validate, that the size encoded size of value indeed exactly matches the value of length. How would you tackle this aspect?

The simplest way is to write a function that returns the expected size of a given type and e.g. `assert(length == TLV::size_of_type(r#type))`. This particular thing of determining struct size has been discussed in some other issues/discussions, so you may do a search around.

> Ah and btw how do you model the encoding of bool?

This depends on the file format you are writing. If it is a simple 0/1, Rust comes with `impl From<bool> for {number}` so you can just use `map` to map to the appropriate number type.
